# Exercises II



- Copy the users and roles DataFrames from the examples above.


In [1]:
import pandas as pd
import numpy as np

In [2]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [3]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter


- What is the result of using a right join on the DataFrames?


In [4]:
users.merge(roles, left_on='role_id', right_on='id', how='right', indicator=True)


id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only


- What is the result of using an outer join on the DataFrames?


In [5]:
# Perform an outer join specifying the left and right DataFrame keys.

users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)


id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only


- What happens if you drop the foreign keys from the DataFrames and try to merge them?


In [6]:
users.merge(roles, how='outer', indicator=True)


id       name  role_id      _merge
0   1        bob      1.0   left_only
1   2        joe      2.0   left_only
2   3      sally      3.0   left_only
3   4       adam      3.0   left_only
4   5       jane      NaN   left_only
5   6       mike      NaN   left_only
6   1      admin      NaN  right_only
7   2     author      NaN  right_only
8   3   reviewer      NaN  right_only
9   4  commenter      NaN  right_only


- Load the mpg dataset from PyDataset.


In [10]:
from pydataset import data
mpg = data('mpg')

In [11]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]


- Output and read the documentation for the mpg dataset.


In [12]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 





- How many rows and columns are in the dataset?


234 rows, 11 columns


- Check out your column names and perform any cleanup you may want on them.


In [15]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [23]:
mpg = mpg.rename(columns={'displ': 'engine_size', 
                     'cyl': 'cylinder',
                     'trans': 'transmission',
                     'drv': 'drive',
                     'cty': 'city',
                     'hwy': 'highway',
                     'fl': 'fuel_type'}
)


- Display the summary statistics for the dataset.


In [24]:
mpg.agg(
    {
            'year': ['min', 'max', 'median'],
            'city': ['min', 'max', 'median'],
            'highway': ['min', 'max', 'median'],
})

year  city  highway
min     1999.0   9.0     12.0
max     2008.0  35.0     44.0
median  2003.5  17.0     24.0


- How many different manufacturers are there?


In [33]:
mpg.value_counts('manufacturer')

manufacturer
dodge         37
toyota        34
volkswagen    27
ford          25
chevrolet     19
audi          18
hyundai       14
subaru        14
nissan        13
honda          9
jeep           8
pontiac        5
land rover     4
mercury        4
lincoln        3
dtype: int64


- How many different models are there?


In [32]:
mpg.value_counts('model')

model
caravan 2wd               11
ram 1500 pickup 4wd       10
civic                      9
jetta                      9
mustang                    9
dakota pickup 4wd          9
impreza awd                8
grand cherokee 4wd         8
a4 quattro                 8
camry solara               7
a4                         7
passat                     7
camry                      7
durango 4wd                7
sonata                     7
tiburon                    7
f150 pickup 4wd            7
toyota tacoma 4wd          7
new beetle                 6
4runner 4wd                6
forester awd               6
explorer 4wd               6
altima                     6
gti                        5
c1500 suburban 2wd         5
corolla                    5
corvette                   5
grand prix                 5
malibu                     5
mountaineer 4wd            4
pathfinder 4wd             4
range rover                4
k1500 tahoe 4wd            4
expedition 2wd             3
maxima  


- Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.


In [35]:
mpg['mileage_difference'] = mpg.highway - mpg.city
mpg

manufacturer   model  engine_size  year  cylinder transmission drive  \
1           audi      a4          1.8  1999         4     auto(l5)     f   
2           audi      a4          1.8  1999         4   manual(m5)     f   
3           audi      a4          2.0  2008         4   manual(m6)     f   
4           audi      a4          2.0  2008         4     auto(av)     f   
5           audi      a4          2.8  1999         6     auto(l5)     f   
..           ...     ...          ...   ...       ...          ...   ...   
230   volkswagen  passat          2.0  2008         4     auto(s6)     f   
231   volkswagen  passat          2.0  2008         4   manual(m6)     f   
232   volkswagen  passat          2.8  1999         6     auto(l5)     f   
233   volkswagen  passat          2.8  1999         6   manual(m5)     f   
234   volkswagen  passat          3.6  2008         6     auto(s6)     f   

     city  highway fuel_type    class  mileage_difference  
1      18       29         p  compact                  11  
2      21       29         p  compact                   8  
3      20       31         p  compact                  11  
4      21       30         p  compact                   9  
5      16       26         p  compact                  10  
..    ...      ...       ...      ...                 ...  
230    19       28         p  midsize                   9  
231    21       29         p  midsize                   8  
232    16       26         p  midsize                  10  
233    18       26         p  midsize                   8  
234    17       26         p  midsize                   9  

[234 rows x 12 columns]


- Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.


In [36]:
mpg['average_mileage'] = mpg.city + mpg.highway / 2
mpg

manufacturer   model  engine_size  year  cylinder transmission drive  \
1           audi      a4          1.8  1999         4     auto(l5)     f   
2           audi      a4          1.8  1999         4   manual(m5)     f   
3           audi      a4          2.0  2008         4   manual(m6)     f   
4           audi      a4          2.0  2008         4     auto(av)     f   
5           audi      a4          2.8  1999         6     auto(l5)     f   
..           ...     ...          ...   ...       ...          ...   ...   
230   volkswagen  passat          2.0  2008         4     auto(s6)     f   
231   volkswagen  passat          2.0  2008         4   manual(m6)     f   
232   volkswagen  passat          2.8  1999         6     auto(l5)     f   
233   volkswagen  passat          2.8  1999         6   manual(m5)     f   
234   volkswagen  passat          3.6  2008         6     auto(s6)     f   

     city  highway fuel_type    class  mileage_difference  average_mileage  
1      18       29         p  compact                  11             32.5  
2      21       29         p  compact                   8             35.5  
3      20       31         p  compact                  11             35.5  
4      21       30         p  compact                   9             36.0  
5      16       26         p  compact                  10             29.0  
..    ...      ...       ...      ...                 ...              ...  
230    19       28         p  midsize                   9             33.0  
231    21       29         p  midsize                   8             35.5  
232    16       26         p  midsize                  10             29.0  
233    18       26         p  midsize                   8             31.0  
234    17       26         p  midsize                   9             30.0  

[234 rows x 13 columns]


- Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.


In [38]:
mpg['is_automatic'] = mpg.transmission('auto')
mpg

TypeError: 'Series' object is not callable


- Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?

- Do automatic or manual cars have better miles per gallon?